stats_test.go

```diff
@@ -55,6 +57,223 @@ func TestTimer(t *testing.T) {
 	}
 }
 
+func TestNewSubScope(t *testing.T) {
+	s := randomTagSet(t, "x_", 20)
+	for i := range s {
+		s[i].value += "|" // add an invalid char
+	}
+	m := make(map[string]string)
+	for _, p := range s {
+		m[p.key] = p.value
+	}
+	scope := newSubScope(nil, "name", m)
+
+	expected := make(tagSet, len(s))
+	copy(expected, s)
+	for i, p := range expected {
+		expected[i].value = replaceChars(p.value)
+	}
+
+	if !reflect.DeepEqual(scope.tags, expected) {
+		t.Errorf("tags are not sorted by key: %+v", s)
+	}
+	for i, p := range expected {
+		s := replaceChars(p.value)
+		if p.value != s {
+			t.Errorf("failed to replace invalid chars: %d: %+v", i, p)
+		}
+	}
+	if scope.name != "name" {
+		t.Errorf("wrong scope name: %s", scope.name)
+	}
+}
+
+func mergeTagsReference(s *subScope, tags map[string]string) tagSet {
+	a := make(tagSet, 0, len(tags))
+	for k, v := range tags {
+		if k != "" && v != "" {
+			a = append(a, tagPair{key: k, value: replaceChars(v)})
+		}
+	}
+	return mergeTagSetsReference(s.tags, a)
+}
+
+func tagMapsEqual(m1, m2 map[string]string) bool {
+	if len(m1) != len(m2) {
+		return false
+	}
+	for k, v := range m1 {
+		if vv, ok := m2[k]; !ok || vv != v {
+			return false
+		}
+	}
+	for k, v := range m2 {
```

Comment:
Might be wrong here but if you already checked the lengths are equal isn't it enough to check that all of m1 is in m2, and you don't need this second check to see if all of m2 is in m1?

Reply:
correct and fixed